### PR TITLE
feat: add field key option to set event partition key

### PR DESCRIPTION
**Pull request by powersj**
This allows users to set the partition key value in data set to Azure.
This value gets hashed and that hash determines which of the available
partitions the event will get sent to. The hash values get split evenly
over the available partitions that exist in Azure.

Fixes: #10762

### CODE REVIEW

**Comment by powersj** [comment]
@jcorioland, @plalwa,

It was easier to create a new PR, so please see the comment above from the telegraf-tiger, which has links to PR artifacts you can download and try it out! Please let me know if something does or does not work :)

Thanks!

**Comment by plalwa** [comment]
> @jcorioland, @plalwa,
> 
> It was easier to create a new PR, so please see the comment above from the telegraf-tiger, which has links to PR artifacts you can download and try it out! Please let me know if something does or does not work :)
> 
> Thanks!

Hi @powersj, I tested it locally the connection works however it wrote to all 3 partitions in round robin instead of 1 partition. I guess though may be coz i am using tag_key like `host` as the partition key and not a filed_key like `cpu_total`. which is currently the requirement to partition based on tag and not field.

**Comment by powersj** [comment]
> I guess though may be coz i am using tag_key like `host` as the partition key and not a filed_key like `n_cpus`. which is currently the requirement to partition based on tag and not field.
> **data:**
> {"fields":{"load1":0,"load15":0,"load5":0,"n_cpus":8},"name":"system","tags":{"host":"localhost"},"timestamp":1652215715}

Right the code as-is will go through the fields, not tags, so it would not set it in this case. I am open to what makes more sense here: if it should look through a tag or a field or maybe both?

**Comment by plalwa** [comment]
> > I guess though may be coz i am using tag_key like `host` as the partition key and not a filed_key like `n_cpus`. which is currently the requirement to partition based on tag and not field.
> > **data:**
> > {"fields":{"load1":0,"load15":0,"load5":0,"n_cpus":8},"name":"system","tags":{"host":"localhost"},"timestamp":1652215715}
> 
> Right the code as-is will go through the fields, not tags, so it would not set it in this case. I am open to what makes more sense here: if it should look through a tag or a field or maybe both?

tag is the currect required for us, however both should be allowed to make it a complete solution.

**Comment by powersj** [comment]
> tag is the currect required for us, however both should be allowed to make it a complete solution.

I've pushed another commit that will check tags first, and then fields. In about 20-30mins the telegraf tiger should add a comment to this PR with artifacts, assuming tests and CI all pass.

Thanks!

**Comment by plalwa** [comment]
> Download PR build artifacts for [linux_amd64.tar.gz](https://output.circle-artifacts.com/output/job/f917ae90-d0e4-4c70-ae42-b1f8839b92df/artifacts/0/build/dist/telegraf-1.23.0~f0da5667_linux_amd64.tar.gz), [darwin_amd64.tar.gz](https://output.circle-artifacts.com/output/job/21324305-da61-40f5-9821-1a349b285c60/artifacts/0/build/dist/telegraf-1.23.0~f39666dc_darwin_amd64.tar.gz), and [windows_amd64.zip](https://output.circle-artifacts.com/output/job/2bdf1471-51d1-4d67-9fc8-b20f8e0a8d18/artifacts/0/build/dist/telegraf-1.23.0~dfd1effb_windows_amd64.zip). Downloads for additional architectures and packages are available below.
> 
> 👍 This pull request doesn't change the Telegraf binary size
> 
> 📦 Click here to get additional PR build artifacts

This works fine as expected, i used tag_key `host` and `cpu` on multiple runs and could see the data got partition based on the key values.
**parition0**
x-opt-partition-keycpu2
**partition1**
x-opt-partition-keycpu7

to me this serves the purpose of requirement, however i would do few more tests.. just to avoid  i miss something,otherwise good to merge :) .

**Comment by powersj** [comment]
> to me this serves the purpose of requirement, however i would do few more tests.. just to avoid i miss something,otherwise good to merge :) .

Thanks for the response and let us know what the rest of your tests discover. Thanks again!

**Comment by plalwa** [comment]
> > Download PR build artifacts for [linux_amd64.tar.gz](https://output.circle-artifacts.com/output/job/f917ae90-d0e4-4c70-ae42-b1f8839b92df/artifacts/0/build/dist/telegraf-1.23.0~f0da5667_linux_amd64.tar.gz), [darwin_amd64.tar.gz](https://output.circle-artifacts.com/output/job/21324305-da61-40f5-9821-1a349b285c60/artifacts/0/build/dist/telegraf-1.23.0~f39666dc_darwin_amd64.tar.gz), and [windows_amd64.zip](https://output.circle-artifacts.com/output/job/2bdf1471-51d1-4d67-9fc8-b20f8e0a8d18/artifacts/0/build/dist/telegraf-1.23.0~dfd1effb_windows_amd64.zip). Downloads for additional architectures and packages are available below.
> > 👍 This pull request doesn't change the Telegraf binary size
> > 📦 Click here to get additional PR build artifacts
> 
> This works fine as expected, i used tag_key `host` and `cpu` on multiple runs and could see the data got partition based on the key values. **parition0** x-opt-partition-key��cpu2� **partition1** x-opt-partition-key��cpu7�
> 
> to me this serves the purpose of requirement, however i would do few more tests.. just to avoid i miss something,otherwise good to merge :) .

had multiple testing on different attributes, looks good, @powersj when can we merge it to get an office version?

File: plugins/outputs/event_hubs/README.md [hunk 1]
```diff
@@ -21,8 +21,15 @@ JSON is probably the easiest to integrate with downstream components.
...
   ## The shared access key must have "Send" permissions on the target Event Hub.
   connection_string = "Endpoint=sb://namespace.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=superSecret1234=;EntityPath=hubName"
+
   ## Client timeout (defaults to 30s)
   # timeout = "30s"
+
+  ## Partition key
+  ## Metric tag or field name to use for the event partition key. The value of
+  ## this tag or field is set as the key for events if it exists.
+  # partition_key = ""
```
Comment (srebhan) [inline plugins/outputs/event_hubs/README.md:31]: Can you please mention here that "If both, tag and field, exists the tag is preferred"?

**Review by srebhan — COMMENTED** [review]
@powersj thanks for the PR! Can you please add a short comment to the sample-config about the preference of tags over fields?

File: plugins/outputs/event_hubs/README.md (hunk 1 above)
Comment (powersj) [reply plugins/outputs/event_hubs/README.md]: done!

**Review by srebhan — APPROVED** [review]
Looks good to me. Thanks!

**Review by sspaink — APPROVED** [review]

**Pull request merged**